# 01 - Data Understanding

## Online Retail Sales & Customer Analysis

This notebook focuses on understanding the structure, content, quality, and business meaning of the Online dataset before performing data cleaning and exploratory analysis.

## 1. Project Objective

The objective of this project is to analyze transaction data from an international online retailer to understand:

- sales and revenue performance
- product performance
- customer perchasing behaviour
- geograohical patterns
- customer value and segmentation

Before performing analysis, we first need to understand the dataset, its variables, its structure, and potential data quality issues.

In [2]:
import pandas as pd

# Load the dataset
df = pd.read_excel("../OnlineRetail.xlsx")

## 2. First Look at the Dataset

We begin by inspecting a few records to understand what one row represents and what information is availables for each transaction.

In [3]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [4]:
df.shape

(541909, 8)

In [5]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='str')

## 3. Data Structure

We inspect the dataset's schema, including the number of observations, data types, and missing values.

In [11]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[us]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 33.1+ MB


In [7]:
df.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

### Initital observations

- The dataset contains 541,909 transactions.
- TThere are 8 columns.
- 'CustomerID' contains missing values.
- 'Description' contains missing values.
- 'InvoiceDate' is already stored as a datetime type.
- 'Quantity' is integer-valued.
- 'UnitePrice' is floating-point.

## 4. Descriptive Statistics

Descriptive statistics provide an initial view of the numerical variables and help identify unusual values or distributions that require further investigation.

In [14]:
df.describe()

,Quantity,InvoiceDate,UnitPrice,CustomerID
count,541909.000000,541909,541909.000000,406829.000000
mean,9.552250,2011-07-04 13:34:57.156386,4.611114,15287.690570
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.081158,NaN,96.759853,1713.600303


## 5. Understanding the Dataset's Business Entities

The dataset contains information at the transaction-line level.
We therefore examine the number of unique invoices, customers, products,and countries.

This helps distinguish between transaction records and higher-level business entities such as orders and customers.

In [16]:
print("Unique invoices:", df["InvoiceNo"].nunique())
print("Unique customers:", df["CustomerID"].nunique())
print("Unique products:", df["StockCode"].nunique())
print("Unique countries:", df["Country"].nunique())

Unique invoices: 25900
Unique customers: 4372
Unique products: 4070
Unique countries: 38


In [30]:
print(df["InvoiceNo"].value_counts().head(10))

InvoiceNo
573585    1114
581219     749
581492     731
580729     721
558475     705
579777     687
581217     676
537434     675
580730     662
538071     652
Name: count, dtype: int64


In [32]:
df["CustomerID"].value_counts().head(10)

CustomerID
17841.0    7983
14911.0    5903
14096.0    5128
12748.0    4642
14606.0    2782
15311.0    2491
14646.0    2085
13089.0    1857
13263.0    1677
14298.0    1640
Name: count, dtype: int64

In [39]:
df["Description"].value_counts().head(10)

Description
WHITE HANGING HEART T-LIGHT HOLDER    2369
REGENCY CAKESTAND 3 TIER              2200
JUMBO BAG RED RETROSPOT               2159
PARTY BUNTING                         1727
LUNCH BAG RED RETROSPOT               1638
ASSORTED COLOUR BIRD ORNAMENT         1501
SET OF 3 CAKE TINS PANTRY DESIGN      1473
PACK OF 72 RETROSPOT CAKE CASES       1385
LUNCH BAG  BLACK SKULL.               1350
NATURAL SLATE HEART CHALKBOARD        1280
Name: count, dtype: int64

In [34]:
df["StockCode"].value_counts().head(10)

StockCode
85123A    2313
22423     2203
85099B    2159
47566     1727
20725     1639
84879     1502
22720     1477
22197     1476
21212     1385
20727     1350
Name: count, dtype: int64

## 6. Geographic Coverage

We examine the countries represented in the dataset to determine
whether the retailer operates only in its domestic UK market or
serves an international customer base.

In [42]:
df["Country"].unique()

<StringArray>
[      'United Kingdom',               'France',            'Australia',
          'Netherlands',              'Germany',               'Norway',
                 'EIRE',          'Switzerland',                'Spain',
               'Poland',             'Portugal',                'Italy',
              'Belgium',            'Lithuania',                'Japan',
              'Iceland',      'Channel Islands',              'Denmark',
               'Cyprus',               'Sweden',              'Austria',
               'Israel',              'Finland',              'Bahrain',
               'Greece',            'Hong Kong',            'Singapore',
              'Lebanon', 'United Arab Emirates',         'Saudi Arabia',
       'Czech Republic',               'Canada',          'Unspecified',
               'Brazil',                  'USA',   'European Community',
                'Malta',                  'RSA']
Length: 38, dtype: str

In [ ]:
df["Country"].value_counts()

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


## 7. Initial Data Quality Investigation

### 7.1 Negative Quantities

The `Quantity` variable contains negative values. Since quantity would
normally represent the number of units purchased, these observations
require investigation before deciding how they should be treated.

In [19]:
df["Quantity"].describe()

count    541909.000000
mean          9.552250
std         218.081158
min      -80995.000000
25%           1.000000
50%           3.000000
75%          10.000000
max       80995.000000
Name: Quantity, dtype: float64

In [29]:
print((df["Quantity"] < 0).sum())

10624


In [22]:
df[df["Quantity"] < 0].count()

InvoiceNo      10624
StockCode      10624
Description     9762
Quantity       10624
InvoiceDate    10624
UnitPrice      10624
CustomerID      8905
Country        10624
dtype: int64

In [20]:
df[df["Quantity"] < 0].head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom


In [23]:
df["InvoiceNo"].astype(str).str.startswith("C").value_counts()

InvoiceNo
False    532621
True       9288
Name: count, dtype: int64

In [24]:
df.loc[
    df["InvoiceNo"].astype(str).str.startswith("C"),
    "Quantity"
].describe()

count     9288.000000
mean       -29.885228
std       1145.786965
min     -80995.000000
25%         -6.000000
50%         -2.000000
75%         -1.000000
max         -1.000000
Name: Quantity, dtype: float64

In [36]:
print("Unique invoices:", df["InvoiceNo"].nunique())

Unique invoices: 25900


In [37]:
negative_quantity = df["Quantity"] < 0
cancelled_invoice = df["InvoiceNo"].astype(str).str.startswith("C")

print("Negative quantity:", negative_quantity.sum())
print("C invoices:", cancelled_invoice.sum())
print("Both:", (negative_quantity & cancelled_invoice).sum())

Negative quantity: 10624
C invoices: 9288
Both: 9288


In [47]:
df.loc[
    (df["Quantity"] < 0) &
    (~df["InvoiceNo"].astype(str).str.startswith("C")),
    ["InvoiceNo", "StockCode", "Description", "Quantity", "UnitPrice", "CustomerID", "Country"]
].head(20)

,InvoiceNo,StockCode,Description,Quantity,UnitPrice,CustomerID,Country
2406,536589,21777,NaN,-10,0.0,NaN,United Kingdom
4347,536764,84952C,NaN,-38,0.0,NaN,United Kingdom
7188,536996,22712,NaN,-20,0.0,NaN,United Kingdom
7189,536997,22028,NaN,-20,0.0,NaN,United Kingdom
7190,536998,85067,NaN,-6,0.0,NaN,United Kingdom
7192,537000,21414,NaN,-22,0.0,NaN,United Kingdom
7193,537001,21653,NaN,-6,0.0,NaN,United Kingdom
7195,537003,85126,NaN,-2,0.0,NaN,United Kingdom
7196,537004,21814,NaN,-30,0.0,NaN,United Kingdom
7197,537005,21692,NaN,-70,0.0,NaN,United Kingdom


### 7.2 Unit Price

The `UnitPrice` variable contains zero, negative, and extreme values.
These observations need to be investigated to determine whether they
represent valid business transactions, special cases, or data quality
issues.

In [25]:
df["UnitPrice"].describe()

count    541909.000000
mean          4.611114
std          96.759853
min      -11062.060000
25%           1.250000
50%           2.080000
75%           4.130000
max       38970.000000
Name: UnitPrice, dtype: float64

In [26]:
print("Zero prices:", (df["UnitPrice"] == 0).sum())
print("Negative prices:", (df["UnitPrice"] < 0).sum())
print("Prices above 100:", (df["UnitPrice"] > 100).sum())

Zero prices: 2515
Negative prices: 2
Prices above 100: 1036


In [51]:
df[df["UnitPrice"] < 0]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


In [52]:
df[df["StockCode"] == "B"]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299982,A563185,B,Adjust bad debt,1,2011-08-12 14:50:00,11062.06,NaN,United Kingdom
299983,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


In [67]:
df[df["UnitPrice"] == 0]["Description"].unique()

array([nan, 'amazon', '?', 'ROUND CAKE TIN VINTAGE GREEN', 'check',
       'damages', 'CREAM SWEETHEART LETTER RACK',
       'ZINC WILLIE WINKIE  CANDLE STICK', 'BOX OF 24 COCKTAIL PARASOLS',
       'DOORMAT ENGLISH ROSE ', 'DOORMAT 3 SMILEY CATS',
       'GREEN REGENCY TEACUP AND SAUCER', 'FRENCH BLUE METAL DOOR SIGN 7',
       'FRENCH BLUE METAL DOOR SIGN 5', 'FRENCH BLUE METAL DOOR SIGN 6',
       'FRENCH BLUE METAL DOOR SIGN 4', 'FRENCH BLUE METAL DOOR SIGN No',
       'FRENCH BLUE METAL DOOR SIGN 8', 'FRENCH BLUE METAL DOOR SIGN 1',
       'RED KITCHEN SCALES', 'IVORY KITCHEN SCALES',
       'SET OF 6 SOLDIER SKITTLES', 'CHILDS GARDEN TROWEL BLUE ',
       'CHILDRENS GARDEN GLOVES BLUE', 'PICNIC BASKET WICKER SMALL',
       'PICNIC BASKET WICKER LARGE', 'EMPIRE UNION JACK TV DINNER TRAY',
       'TV DINNER TRAY VINTAGE PAISLEY', 'SPACEBOY TV DINNER TRAY',
       'TV DINNER TRAY DOLLY GIRL', 'CHILDS GARDEN SPADE BLUE',
       'CHILDS GARDEN RAKE BLUE', 'WATERING CAN PINK BUNNY',
  

In [65]:
df[df["UnitPrice"] == 0]["CustomerID"].value_counts()

CustomerID
13081.0    4
14646.0    4
14911.0    2
12415.0    2
13985.0    2
12647.0    1
16560.0    1
15107.0    1
17560.0    1
13239.0    1
13113.0    1
14410.0    1
12457.0    1
17667.0    1
16818.0    1
12507.0    1
15581.0    1
16133.0    1
12748.0    1
15804.0    1
12446.0    1
18059.0    1
14110.0    1
12437.0    1
12431.0    1
13014.0    1
15602.0    1
12603.0    1
12444.0    1
16406.0    1
13256.0    1
Name: count, dtype: int64

In [35]:
df[df["UnitPrice"] > 100][
    [
        "InvoiceNo",
        "StockCode",
        "Description",
        "Quantity",
        "UnitPrice",
        "CustomerID",
        "Country"
    ]
].head(20)

,InvoiceNo,StockCode,Description,Quantity,UnitPrice,CustomerID,Country
246,536392,22827,RUSTIC SEVENTEEN DRAWER SIDEBOARD,1,165.00,13705.0,United Kingdom
1814,536544,DOT,DOTCOM POSTAGE,1,569.77,NaN,United Kingdom
3041,536592,DOT,DOTCOM POSTAGE,1,607.49,NaN,United Kingdom
4989,536835,22655,VINTAGE RED KITCHEN CABINET,1,295.00,13145.0,United Kingdom
5450,536862,DOT,DOTCOM POSTAGE,1,254.43,NaN,United Kingdom
5545,536864,DOT,DOTCOM POSTAGE,1,121.06,NaN,United Kingdom
5685,536865,DOT,DOTCOM POSTAGE,1,498.47,NaN,United Kingdom
6165,536876,DOT,DOTCOM POSTAGE,1,887.52,NaN,United Kingdom
10812,537237,DOT,DOTCOM POSTAGE,1,863.74,NaN,United Kingdom
11381,537240,DOT,DOTCOM POSTAGE,1,940.87,NaN,United Kingdom


### 7.3 Missing CustomerID



In [77]:
df[df["CustomerID"].isna()]["Description"].value_counts().head(20)

Description
DOTCOM POSTAGE                        693
JUMBO BAG RED RETROSPOT               497
JUMBO STORAGE BAG SUKI                414
JUMBO SHOPPER VINTAGE RED PAISLEY     388
JUMBO BAG WOODLAND ANIMALS            372
JUMBO BAG PINK POLKADOT               348
RECYCLING BAG RETROSPOT               341
RED TOADSTOOL LED NIGHT LIGHT         328
SUKI  SHOULDER BAG                    326
GREEN REGENCY TEACUP AND SAUCER       324
PARTY BUNTING                         311
PACK OF 72 RETROSPOT CAKE CASES       305
WOODLAND CHARLOTTE BAG                304
RED RETROSPOT CHARLOTTE BAG           301
CHARLOTTE BAG SUKI DESIGN             300
WHITE HANGING HEART T-LIGHT HOLDER    299
REGENCY CAKESTAND 3 TIER              295
JAM MAKING SET PRINTED                294
RECIPE BOX PANTRY YELLOW DESIGN       290
JUMBO  BAG BAROQUE BLACK WHITE        287
Name: count, dtype: int64

## 8. Initial Findings

Based on the initial investigation, we identified several areas
requiring further investigation before data cleaning:

### 8.1 Business Entities

- The dataset contains 541,909 transaction lines, representing 25,900 unique invoices.

- There are 4,372 unique identified customers and 4,070 unique products across 38 countries.

- The number of transaction lines is much larger than the number of invoices because a single invoice can contain multiple product lines.

- CustomerID is missing for 135,080 transaction lines, which means customer-level analysis may require careful treatment of transactions without an identified customer.

- The high frequency of some InvoiceNo and StockCode values also confirms that the dataset is structured at the transaction-line level rather than at the order or customer level.

### 8.2 Negative Quantities

- The dataset contains 10,624 transaction lines with negative quantities.

- Among these, 9,288 rows have an `InvoiceNo` beginning with `C`. All 9,288 C-prefixed invoice rows have negative quantities, indicating a strong relationship between the `C` prefix and negative-quantity transactions.

- However, 1,336 negative-quantity rows do not have a C prefix. Therefore, the relationship is not one-to-one.

- At this stage, negative quantities should not be removed automatically. Their business meaning and appropriate treatment will be determined during the data-cleaning stage based on the project's analytical objectives.

### 8.3 Unit Price

- 2,515 rows have UnitPrice = 0
- 2 rows have negative UnitPrice, both "Adjust bad debt"
- 1,036 rows have UnitPrice > 100, but these include both expensive products and special transactions such as postage, fees, discounts, and manual entries
- High UnitPrice values cannot be classified as errors simply because they are high. They include legitimate expensive products as well as postage, fees, discounts, and manual transactions.

## 9. Open Questions Before Data Cleaning

Before modifying the dataset, we need to determine:

1. What do negative quantities represent?
2. What do zero and negative unit prices represent?
3. How should cancellation transactions be treated?
4. How should missing CustomerID values affect customer-level analysis?
5. Are there duplicate transaction records?
6. Are extreme values legitimate business transactions?